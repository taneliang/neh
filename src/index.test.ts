import { CloudflareWorkerGlobalScope } from 'types-cloudflare-worker';
declare const self: CloudflareWorkerGlobalScope;

import makeCloudflareWorkerEnv, { makeCloudflareWorkerRequest } from 'cloudflare-worker-mock';

async function getResponse(path: string): Promise<Response> {
  const request = makeCloudflareWorkerRequest(path);
  const responses = await self.trigger('fetch', request);
  return responses[0];
}

describe('neh', () => {
  beforeEach(() => {
    // Merge the Cloudflare Worker Environment into the global scope.
    Object.assign(global, makeCloudflareWorkerEnv());
    // Clear all module imports.
    jest.resetModules();
    // Import and init the Worker.
    jest.requireActual('.');
  });

  // Ensure that test env has been set up correctly.
  test('should add listeners', async () => {
    expect(self.listeners.get('fetch')).toBeDefined();
  });

  test('should respond to OpenSearch endpoint', async () => {
    const response = await getResponse('/_opensearch');
    expect(response.status).toBe(200);
    const responseBody = await response.text();
    expect(responseBody).toContain('<?xml');
    expect(responseBody).toContain('OpenSearchDescription');
  });

  // TODO: Move to handler tests
  test('should respond to root with list', async () => {
    const rootResponse = await getResponse('/');
    const listResponse = await getResponse('/list');
    expect(rootResponse).toEqual(listResponse);
  });

  // TODO: Move to handler tests
  test('should default to DuckDuckGo', async () => {
    const noCommandResponse = await getResponse('search query');
    const withCommandResponse = await getResponse('d search query');
    expect(noCommandResponse).toEqual(withCommandResponse);
  });

  // Some test input
  const testCases = [
    '1+1',
    'drive 2',
    'drive',
    'gh r taneliang/neh',
    'ip',
    "lyrics here's to never growing up",
    'npm p @babel/core',
    'search query',
    'trzh hong kong',
  ];

  testCases.forEach((inputStr) => {
    test(`should handle "${inputStr}" correctly`, async () => {
      const response = await getResponse(inputStr);
      expect(response).toMatchSnapshot();
    });
  });
});
