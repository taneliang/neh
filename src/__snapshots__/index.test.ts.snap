// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`neh should handle "1+1" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://duckduckgo.com/?q=1%2B1",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "drive 2" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://drive.google.com/drive/u/2/",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "drive" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://drive.google.com",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "gh r taneliang/neh" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://github.com/taneliang/neh",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "ip" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://icanhazip.com",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "lyrics here's to never growing up" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://duckduckgo.com/?q=%5Csite:genius.com%20here's%20to%20never%20growing%20up",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "npm p @babel/core" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://www.npmjs.com/package/@babel/core",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "search query" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://duckduckgo.com/?q=search+query",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;

exports[`neh should handle "trzh hong kong" correctly 1`] = `
Response {
  "body": null,
  "bodyUsed": false,
  "headers": Headers {
    "_map": Map {
      "location" => "https://fanyi.baidu.com/#en/zh/hong%20kong",
    },
  },
  "method": "GET",
  "ok": false,
  "redirected": false,
  "status": 302,
  "statusText": "OK",
  "type": "basic",
  "url": "http://example.com/asset",
}
`;
